account-kit/smart-contracts/src/ma-v2/account/common/modularAccountV2Base.ts

```diff
@@ -0,0 +1,217 @@
+import {
+  createBundlerClient,
+  getEntryPoint,
+  InvalidEntityIdError,
+  InvalidNonceKeyError,
+  type AccountOp,
+  type EntryPointDef,
+  type SmartAccountSigner,
+  type SmartContractAccountWithSigner,
+  type ToSmartContractAccountParams,
+} from "@aa-sdk/core";
+import { DEFAULT_OWNER_ENTITY_ID } from "../../utils.js";
+import {
+  type Hex,
+  type Address,
+  type Chain,
+  type Transport,
+  encodeFunctionData,
+  maxUint32,
+  zeroAddress,
+  getContract,
+  concatHex,
+  maxUint152,
+} from "viem";
+import { modularAccountAbi } from "../../abis/modularAccountAbi.js";
+import { serializeModuleEntity } from "../../actions/common/utils.js";
+
+export const executeUserOpSelector: Hex = "0x8DD7712F";
+
+export type SignerEntity = {
+  isGlobalValidation: boolean;
+  entityId: number;
+};
+
+export type ExecutionDataView = {
+  module: Address;
+  skipRuntimeValidation: boolean;
+  allowGlobalValidation: boolean;
+  executionHooks: readonly Hex[];
+};
+
+export type ValidationDataView = {
+  validationHooks: readonly Hex[];
+  executionHooks: readonly Hex[];
+  selectors: readonly Hex[];
+  validationFlags: number;
+};
+
+export type ValidationDataParams =
+  | {
+      validationModuleAddress: Address;
+      entityId?: never;
+    }
+  | {
+      validationModuleAddress?: never;
+      entityId: number;
+    };
+
+export type MAV2Account<
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = SmartContractAccountWithSigner<"MAV2Account", TSigner, "0.7.0"> & {
+  signerEntity: SignerEntity;
+  getExecutionData: (selector: Hex) => Promise<ExecutionDataView>;
+  getValidationData: (
+    args: ValidationDataParams
+  ) => Promise<ValidationDataView>;
+  encodeCallData: (callData: Hex) => Promise<Hex>;
+};
+
+export type CreateMAV2BaseFunctionsParams<
+  TTransport extends Transport = Transport
+> = Pick<
+  ToSmartContractAccountParams<"MAV2Account", TTransport, Chain, "0.7.0">,
+  "transport" | "chain"
+> & {
+  // salt?: bigint;
+  // factoryAddress?: Address;
+  // initCode?: Hex;
+  // initialOwner?: Address;
+  entryPoint?: EntryPointDef<"0.7.0", Chain>;
+  signerEntity?: SignerEntity;
+  accountAddress: Address;
+};
+
+export async function createMAv2BaseFunctions(
+  config: CreateMAV2BaseFunctionsParams
+) {
+  const {
+    transport,
+    chain,
+    entryPoint = getEntryPoint(chain, { version: "0.7.0" }),
+    signerEntity: {
+      isGlobalValidation = true,
+      entityId = DEFAULT_OWNER_ENTITY_ID,
+    } = {},
```

Comment:
oh holy shit... this is what you were talking about hahaha I remember seeing this during hackathon and being like wtf is going on here. Yea I concede... destructuring CAN go too far XD

Reply:
This isn't even the worst of it 😂 , the messy one is in the old SMA client deleted by this PR (due to refactoring out a common base). The messy one had a destructure for both `signerEntity` as an object, and the inner fields within it, all with default initializers.